src/gstreamer_bridge.cpp

```diff
@@ -26,61 +62,65 @@ GstreamerRosBridge::GstreamerRosBridge(ros::NodeHandle &nh)
     nh.getParam("mtu", mtu);
+    ros_rate_ = ros::Rate(gst_fps);  // update ros rate
 
     setenv("GST_DEBUG", "3", 1); // somehow this helps with querying video position
-    std::ostringstream pipelineStream;
-    pipelineStream 
+
+    std::ostringstream udpPipeline;     // gstreamer pipeline for udp to peer
+    udpPipeline 
         << "appsrc ! videoconvert ! videoscale !"
         << " video/x-raw,width=" << gst_width_ 
         << ",height=" << gst_height_
         << ",framerate=" << gst_fps << "/1 !"
         << " x264enc bitrate=" << bitrate
         << " speed-preset=ultrafast tune=zerolatency !"
-        << " rtph264pay mtu=900 ! udpsink host=" << gs_ip
+        << " rtph264pay mtu=" << mtu << " !"
+        << " udpsink host=" << gs_ip
         << " port=" << gs_port
         << " sync=false";
+    
+    std::string gst_topic = "/camera/image_rect";
+    nh.getParam("gst_topic", gst_topic);
 
-    std::string gstreamer_pipeline = pipelineStream.str();
-    nh.getParam("custom_pipeline", gstreamer_pipeline);
-        
+    // start the udp pipeline
+    std::string gstreamer_pipeline = udpPipeline.str();
     pipeline_.open(gstreamer_pipeline, cv::CAP_GSTREAMER, 0, gst_fps, cv::Size(gst_width_, gst_height_), true);
-    //pipeline_.open(gstreamer_pipeline, cv::CAP_GSTREAMER, true);
     if (!pipeline_.isOpened())
     {
         ROS_ERROR("Failed to open gstreamer pipeline");
         return;
     }
 
-    std::string gst_topic = "/camera/image_rect";
-    nh.getParam("gst_topic", gst_topic);
-
-    gsImageSub_ = nh.subscribe(gst_topic, 1, &GstreamerRosBridge::gsImageCallback, this);
-    rosCameraInfoPub_ = nh.advertise<sensor_msgs::CameraInfo>("/camera/camera_info", 1);
-    rosImagePub_ = nh.advertise<sensor_msgs::Image>("/camera/image_rect", 1);
+    // ros subscribers and publishers
+    gsImageSub_ = nh.subscribe(gst_topic, 1, &GStreamerRosBridge::GsImageCallback, this);
 }
 
-GstreamerRosBridge::~GstreamerRosBridge()
+GStreamerRosBridge::~GStreamerRosBridge()
 {
-    cap_.release();
     pipeline_.release();
 }
 
-
-void GstreamerRosBridge::gsImageCallback(const sensor_msgs::ImageConstPtr &msg)
+void GStreamerRosBridge::GsImageCallback(const sensor_msgs::ImageConstPtr &msg)
 {
+    if (!pipeline_.isOpened()) {
+        ROS_ERROR("GStreamer pipeline is not opened, unable to write image");
+        return;
+    }
+    if (!msg) {
+        ROS_ERROR("Received an empty image message.");
+        return;
+    }   
     // write to gstreamer pipeline
+    cv::Mat image;
     try {
         cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
-        cv::Mat image = cv_ptr->image;
-        cv::resize(image, image, cv::Size(gst_width_, gst_height_));
-        image.convertTo(image, CV_8UC3);
-        pipeline_.write(image);
+        image = cv_ptr->image;
     } catch (cv_bridge::Exception &e) {
         ROS_ERROR("cv_bridge exception: %s", e.what());
     }
+    image.convertTo(image, CV_8UC3);
```

Comment:
I just realized, is this operation necessary? Wouldn't `cv_bridge::toCvCopy(msg, "bgr8")` already give you a 3-channel image with each one being an 8-bit unsigned int? Would this just introduce unnecessary overhead?

Reply:
yes I was just a bit paranoid due to a couple of errors, ill fix that